src/dbSta/src/dbNetwork.cc

```diff
@@ -598,60 +681,112 @@ const char* dbNetwork::name(const Instance* instance) const
+  }
+  return nullptr;
+}
 
-  std::map<std::string, dbModBTerm*> name2modbterm;
+////////////////////////////////////////////////////////////////
+// Module port iterator, allows traversal across dbModulePorts
+// Traverse the ports in creation order (from end to beginning).
 
-  for (auto modbterm : master->getModBTerms()) {
-    const char* port_name = modbterm->getName();
-    Port* port = ConcreteNetwork::makePort(local_cell, port_name);
-    PortDirection* dir = dbToSta(modbterm->getSigType(), modbterm->getIoType());
-    setDirection(port, dir);
-    name2modbterm[std::string(port_name)] = modbterm;
+class dbModulePortIterator : public CellPortIterator
+{
+ public:
+  explicit dbModulePortIterator(const dbModule* cell);
+  ~dbModulePortIterator() override;
+  virtual bool hasNext() override;
+  virtual Port* next() override;
+
+ private:
+  const dbModBTerm* iter_;
+  const dbModule* module_;
+  int ix;
+};
+
+dbModulePortIterator::dbModulePortIterator(const dbModule* cell)
+{
+  // skip to end
+  const dbModBTerm* first_mod_bterm = cell->getHeadDbModBTerm();
+  for (iter_ = cell->getHeadDbModBTerm(); iter_; iter_ = iter_->getNext()) {
+    first_mod_bterm = iter_;
   }
+  iter_ = first_mod_bterm;
+  ix = 0;
+  module_ = cell;
+}
 
-  // make the bus ports. This will generate the bus bits.
-  groupBusPorts(local_cell, [=](const char* port_name) {
-    return portMsbFirst(port_name, master->getName());
-  });
+dbModulePortIterator::~dbModulePortIterator()
+{
+}
 
-  CellPortIterator* ccport_iter = portIterator(local_cell);
-  while (ccport_iter->hasNext()) {
-    Port* cport = ccport_iter->next();
-    const ConcretePort* ccport = reinterpret_cast<const ConcretePort*>(cport);
-    std::string port_name = ccport->name();
-
-    if (ccport->isBus()) {
-      PortMemberIterator* pmi = memberIterator(cport);
-      while (pmi->hasNext()) {
-        Port* bitport = pmi->next();
-        const ConcretePort* cbitport
-            = reinterpret_cast<const ConcretePort*>(bitport);
-        dbModBTerm* modbterm = name2modbterm[std::string(cbitport->name())];
-        modbterm->staSetPort(bitport);
-      }
-    } else if (ccport->isBundle()) {
-      ;
-    } else if (ccport->isBusBit()) {
-      ;
-    } else {
-      dbModBTerm* modbterm = name2modbterm[port_name];
-      modbterm->staSetPort(cport);
+bool dbModulePortIterator::hasNext()
+{
+  if (iter_) {
+    return true;
+  }
+  return false;
+}
+
+Port* dbModulePortIterator::next()
+{
+  ix++;
+  dbModBTerm* modbterm = const_cast<dbModBTerm*>(iter_);
+  Port* ret = reinterpret_cast<Port*>(modbterm);
+  // advance to next, in case of bus port
+  // next modbterm is the one after the bus port
+  if (modbterm->isBusPort()) {
+    dbBusPort* bp = modbterm->getBusPort();
+    int size = bp->getSize();
+    // skip into the content of the bus
+    iter_ = iter_->getPrev();
+    for (int skip_ix = 0; skip_ix < size && (iter_->getPrev()); skip_ix++) {
+      iter_ = iter_->getPrev();
```

Comment:
Do we need prev here or could we have just skipped forward?

Reply:
Yes we can just move forward if the list is reversed (a consequence of the pre-pending rather than appending)

Reply:
removed. Hidden the list traversal behind a custom iterator.